athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/common/PartitionUtil.java

```diff
@@ -0,0 +1,212 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.amazonaws.athena.connectors.gcs.common;
+
+import com.amazonaws.services.glue.model.Column;
+import com.amazonaws.services.glue.model.Table;
+import org.apache.arrow.vector.complex.reader.FieldReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.amazonaws.athena.connectors.gcs.GcsConstants.PARTITION_PATTERN_KEY;
+
+public class PartitionUtil
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(PartitionUtil.class);
+
+    /**
+     * Pattern from a regular expression that identifies a match in a phrases to see if there is any
+     * partition key variable placeholder. A partition key variable placeholder looks something like the following:
+     * year={year}/month={month}
+     * Here, {year} and {month} are the partition key variable placeholders
+     */
+    private static final Pattern PARTITION_PATTERN = Pattern.compile("\\{(.*?)}");
+
+    /**
+     * Match any alpha-num characters. Used to match VARCHAR type only partition keys
+     */
+    private static final String VARCHAR_OR_STRING_REGEX = "([a-zA-Z0-9]+)";
+
+    private PartitionUtil()
+    {
+    }
+
+    /**
+     * Return a regular expression for partition pattern from AWS Glue. This will dynamically generate a
+     * regular expression to match a folder within the GCS to see if the folder conforms with the partition keys
+     * already setup in the AWS Glue Table (if any)
+     *
+     * @param table response of get table from AWS Glue
+     * @return optional Sting of regular expression
+     */
+    public static Optional<String> getRegExExpression(Table table)
+    {
+        List<Column> partitionColumns = table.getPartitionKeys();
+        validatePartitionColumnTypes(partitionColumns);
+        String partitionPattern = table.getParameters().get(PARTITION_PATTERN_KEY);
+        // Check to see if there is a partition pattern configured for the Table by the user
+        // if not, it returns empty value
+        if (partitionPattern == null || partitionPattern.isBlank()) {
+            return Optional.empty();
+        }
+        Matcher partitionMatcher = PARTITION_PATTERN.matcher(partitionPattern);
+        String folderMatchingRegexPattern = partitionPattern;
+        while (partitionMatcher.find()) {
+            folderMatchingRegexPattern = folderMatchingRegexPattern.replace("{" + partitionMatcher.group(1) + "}", VARCHAR_OR_STRING_REGEX);
+        }
+        return Optional.of(folderMatchingRegexPattern);
```

Comment:
As I mentioned in one of my other comments, can't all of these lines (78-83) just be replaced with:

```
folderMatchingRegexPattern.replaceAll(PARTITION_PATTERN, VARCHAR_OR_STRING_REGEX);
```

Try it out and let me know if I'm wrong or missing something.